### PR TITLE
fiber: add fiber_object:csw() and fiber_object:info() functions

### PR DESCRIPTION
**Pull request by iskander232**
Add `fiber_object:csw()` function to get `csw` from both alive or dead fiber.
Add `fiber_object:info()` function to get `info` from alive fiber.

### CODE REVIEW

**Comment by Gerold103** [comment]
Why would you need csw from a dead fiber?

**Comment by iskander232** [comment]
> Why would you need csw from a dead fiber?

I think it is part of the issue

>Besides, `csw` can't be obtained if the fiber is dead. Even if you have fiber object in Lua still available. It returns `:id()`, `:status()`, but can't return `csw`, and is not available via `.info()` anymore.

**Comment by Gerold103** [comment]
> I think it is part of the issue

But it does not mean you need to blindly do it. I see Mons later said "I think not only csw, but all of the fiber info." which will be hardly possible for dead fibers. Also in the usecase I provided in the issue it was about checking if a yield has happened. It is usually needed inside of the current fiber which can't be dead, otherwise it wouldn't be the current one.

Maybe let it throw for dead fibers? I see no usecase for that, but I see a notable code complication.

**Comment by iskander232** [comment]
> Maybe let it throw for dead fibers? I see no usecase for that, but I see a notable code complication.

I agree with you - this code is more complicated and if there are no usecases for getting csw of dead fibers then it is better to let it throw in this case.

**Review by Korablev77 — COMMENTED** [review]
Missing changelog I guess

File: src/lua/fiber.c [hunk 1]
```diff
@@ -156,11 +156,10 @@ static struct fiber *
 lbox_checkfiber(struct lua_State *L, int index)
 {
 	uint64_t fid;
-	if (lua_type(L, index) == LUA_TNUMBER) {
+	if (lua_type(L, index) == LUA_TNUMBER)
 		fid = luaL_touint64(L, index);
-	} else {
+	else
 		fid = *(uint64_t *)luaL_checkudata(L, index, fiberlib_name);
-	}
```
Comment (Korablev77) [inline src/lua/fiber.c:163]: This codestyle fix doesn't seem to be related to the patch

**Comment by iskander232** [comment]
> Missing changelog I guess

Add changelogs

File: src/lua/fiber.c (hunk 1 above)
Comment (iskander232) [reply src/lua/fiber.c]: Remove codestyle fix

File: changelogs/unreleased/gh-5799-add-csw-to-fiber-object.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## feature/lua/fiber
+
+* Introduce `fiber_object:csw()` to get `csw` from fiber
```
Comment (Gerold103) [inline changelogs/unreleased/gh-5799-add-csw-to-fiber-object.md:3]: Please, add a reference to the ticket in the form `(gh-5799)`. Also you don't need multiple changelog files for one feature.

File: src/lua/fiber.c [hunk 3]
```diff
@@ -527,6 +542,17 @@ lbox_fiber_new(struct lua_State *L)
 	return 1;
 }
 
+struct fiber *
+get_fiber(struct lua_State *L)
```
Comment (Gerold103) [inline src/lua/fiber.c:546]: This must be `static`. Also there is clearly a naming convention in this file, and 2 similar getters: `lbox_checkfiber()` and `lbox_fiber_id()`. I suggest you to conform, and name it similarly. For instance, `lbox_get_fiber()`.

File: src/lua/fiber.c [hunk 4]
```diff
@@ -527,6 +542,17 @@ lbox_fiber_new(struct lua_State *L)
...
+{
+	if (lua_gettop(L)) {
+		uint64_t  *fid = luaL_checkudata(L, 1, fiberlib_name);
```
Comment (Gerold103) [inline src/lua/fiber.c:549]: Double whitespace after the type.

File: src/lua/fiber.c [hunk 5]
```diff
@@ -527,6 +542,17 @@ lbox_fiber_new(struct lua_State *L)
 	return 1;
 }
 
+struct fiber *
+get_fiber(struct lua_State *L)
+{
+	if (lua_gettop(L)) {
```
Comment (Gerold103) [inline src/lua/fiber.c:548]: In the new code we do not use implicit casts between integers and booleans. If you need to check the stack is not empty, you need to do `lua_gettop(L) != 0`.

File: src/lua/fiber.c [hunk 6]
```diff
@@ -563,6 +582,42 @@ lbox_fiber_status(struct lua_State *L)
...
 
+/**
+ * Get fiber info: number of context switches, backtrace, id,
+ * total memory, used memory.
+ */
+static int
+lbox_fiber_object_info(struct lua_State *L)
+{
+	struct fiber *f = get_fiber(L);
+	if (f == NULL || f->fid == 0)
```
Comment (Gerold103) [inline src/lua/fiber.c:593]: How can its fid be zero? Is there a test confirming this check is needed? If this is the self fiber, it is not dead by definition. If it is not self but dead, then `fiber_find()` in `get_fiber()` wouldn't find it, and `f == NULL` would be true. Do I miss something?
I see you don't do the fid check in `lbox_fiber_csw()` even though it is not any different.

File: test/box/gh-5799-fiber-info.result [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+test_run:cmd('setopt delimiter ";"')
+ | ---
+ | - true
+ | ...
+for i=1,100 do
+    fibers[i] = fiber.new(function()
+        for j=1,10 do
+            fiber.yield()
+            if j == fiber.self():csw() and j == fiber.self():info().csw then
```
Comment (Gerold103) [inline test/box/gh-5799-fiber-info.result:31]: Please, keep the lines in 80 symbols length limit. This is clearly longer.

File: test/box/gh-5799-fiber-info.result [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+test_run:cmd('setopt delimiter ""');
+ | ---
+ | - true
+ | ...
+csw_check_counter
+ | ---
+ | - 1000
+ | ...
+
+running = fiber.create(function() fiber.sleep(3) end)
```
Comment (Gerold103) [inline test/box/gh-5799-fiber-info.result:54]: Please, avoid creating long-living fibers. It might live too long and affect some of the next tests making them flaky. Try to use fiber cond or a busy-yield-loop in the fiber's body to wait for its proper termination before the test ends.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: changelogs/unreleased/gh-5799-add-csw-to-fiber-object.md (hunk 2 above)
Comment (iskander232) [reply changelogs/unreleased/gh-5799-add-csw-to-fiber-object.md]: There are 2 features (new functions in `fiber_object` and setting `fiber->csw=0`) in this patch so, I think, there should be 2 changelog files.

Added reference to the ticket and reduced number of changelog files from 3 to 2.

File: src/lua/fiber.c (hunk 3 above)
Comment (iskander232) [reply src/lua/fiber.c]: Fixed

File: src/lua/fiber.c (hunk 4 above)
Comment (iskander232) [reply src/lua/fiber.c]: Fixed

File: src/lua/fiber.c (hunk 5 above)
Comment (iskander232) [reply src/lua/fiber.c]: Fixed

File: src/lua/fiber.c (hunk 6 above)
Comment (iskander232) [reply src/lua/fiber.c]: You are right. Removed extra check.

File: test/box/gh-5799-fiber-info.result (hunk 8 above)
Comment (iskander232) [reply test/box/gh-5799-fiber-info.result]: Added `fiber.cond()` instead of `fiber.sleep()`.

File: test/box/gh-5799-fiber-info.result (hunk 7 above)
Comment (iskander232) [reply test/box/gh-5799-fiber-info.result:31]: Fixed

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the fixes! See the last comment below!

File: src/lua/fiber.c [hunk 9]
```diff
@@ -313,6 +310,16 @@ lbox_fiber_statof(struct fiber *f, void *cb_ctx, bool backtrace)
 		lua_settable(L, -3);
 #endif /* ENABLE_BACKTRACE */
 	}
+	return 0;
+}
+
+static int
+lbox_fiber_statof(struct fiber *f, void *cb_ctx, bool backtrace)
+{
+	struct lua_State *L = (struct lua_State *) cb_ctx;
```
[the remark this answers is not in the document]
Comment (iskander232) [reply src/lua/fiber.c]: Fixed

**Comment by Gerold103** [comment]
Sorry, I didn't notice it before - please, move the test to `test/app`. It is not related to box API anyhow, so it should not be in `test/box`.

**Review by Gerold103 — APPROVED** [review]
This is a "pre-approve". Assuming that you will move the test, and Nikita will take a second look.

**Review by Korablev77 — APPROVED** [review]
